### PR TITLE
zxfer: init at 1.1.7

### LINKED ISSUE #174660: Packaging request: zxfer
**Project description**
A popular script in BSD-land for managing ZFS snapshot replication.
It allows the easy and reliable backup, restore or transfer of ZFS filesystems, either locally or remotely.

**Metadata**

* homepage URL: https://github.com/allanjude/zxfer
* source URL: https://github.com/allanjude/zxfer
* license: bsd-2clause
* platforms: unix

**Package example**

FreeBSD ports: https://www.freshports.org/sysutils/zxfer 
AUR: https://aur.archlinux.org/packages/zxfer

### PR DESCRIPTION
**Pull request by urandom2**
###### Description of changes

Fixes #174660

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/system/zxfer/default.nix [hunk 1]
```diff
@@ -0,0 +1,45 @@
+{ coreutils
+, fetchFromGitHub
+, gawk
+, gnugrep
+, gnused
+, lib
```
Comment (viraptor) [inline pkgs/applications/system/zxfer/default.nix:6]: There's a convention for putting `lib` first on the list, then going in the order they're used.

File: pkgs/applications/system/zxfer/default.nix [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    rev = "v${version}";
+    hash = "sha256-11SQJcD3GqPYBIgaycyKkc62/diVKPuuj2Or97j+NZY=";
+  };
+
+  nativeBuildInputs = [ installShellFiles makeWrapper ];
+
+  installPhase = ''
+    installManPage zxfer.1m zxfer.8
+    install -Dm755 ${pname} -t $out/bin/
+    wrapProgram $out/bin/${pname} --prefix PATH : ${lib.makeBinPath [
```
Comment (SuperSandro2000) [inline pkgs/applications/system/zxfer/default.nix:30]:
```suggestion
    install -Dm755 zxfer -t $out/bin/
    wrapProgram $out/bin/zxfer --prefix PATH : ${lib.makeBinPath [
```

File: pkgs/applications/system/zxfer/default.nix [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      coreutils
+      gawk
+      gnugrep
+      gnused
```
Comment (SuperSandro2000) [inline pkgs/applications/system/zxfer/default.nix:34]:
```suggestion
      gawk
```

Comment (urandom2) [reply pkgs/applications/system/zxfer/default.nix:34]: > waiting on consensus from https://github.com/NixOS/nixpkgs/pull/208333#discussion_r1059488315

File: pkgs/applications/system/zxfer/default.nix [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      rsync
+      zfs
+    ]}
+  '';
+
+  meta = with lib; {
+    description = "A continuation of development on zxfer, a popular script for managing ZFS snapshot replication";
```
Comment (AndersonTorres) [inline pkgs/applications/system/zxfer/default.nix:41]: Short descriptions should not be self-referent

File: pkgs/applications/system/zxfer/default.nix [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  src = fetchFromGitHub {
+    owner = "allanjude";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-11SQJcD3GqPYBIgaycyKkc62/diVKPuuj2Or97j+NZY=";
+  };
+
+  nativeBuildInputs = [ installShellFiles makeWrapper ];
+
+  installPhase = ''
```
Comment (AndersonTorres) [inline pkgs/applications/system/zxfer/default.nix:27]:
```suggestion
  installPhase = ''
    runHook preInstall

```

File: pkgs/applications/system/zxfer/default.nix [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    install -Dm755 zxfer -t $out/bin/
+    wrapProgram $out/bin/zxfer --prefix PATH : ${lib.makeBinPath [
+      coreutils
+      gawk
+      gnugrep
+      gnused
+      rsync
+      zfs
+    ]}
+  '';
```
Comment (AndersonTorres) [inline pkgs/applications/system/zxfer/default.nix:38]:
```suggestion

    runHook postInstall
  '';
```

**Review by AndersonTorres — DISMISSED** [review]

File: pkgs/applications/system/zxfer/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/applications/system/zxfer/default.nix:34]: Its not in the default installed tools

Comment (urandom2) [reply pkgs/applications/system/zxfer/default.nix:36]: can you point me to a listing of these "base installed tools"?

Comment (AndersonTorres) [reply pkgs/applications/system/zxfer/default.nix:36]: These things are usually listed around many NixOS modules...

File: pkgs/applications/system/zxfer/default.nix [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+, zfs
+}:
+
+stdenv.mkDerivation rec {
+  pname = "zxfer";
+  version = "1.1.7";
+
+  src = fetchFromGitHub {
+    owner = "allanjude";
+    repo = pname;
```
Comment (AndersonTorres) [inline pkgs/applications/system/zxfer/default.nix:20]:
```suggestion
    repo = "zxfer";
```

It is not so much useful to parameterize the pname.

Comment (urandom2) [reply pkgs/applications/system/zxfer/default.nix:20]: I actually find it to be very helpful at reducing typos

Comment (urandom2) [reply pkgs/applications/system/zxfer/default.nix:20]: I actually find it to be very helpful at reducing typos

File: pkgs/applications/system/zxfer/default.nix (hunk 3 above)
Comment (urandom2) [reply pkgs/applications/system/zxfer/default.nix]: I have stolen the approach from #208333, so this is probably moot, but I would be keen to get a link to those tools in #209374 as I do not know what to look for in `./nixos`

File: pkgs/applications/system/zxfer/default.nix [hunk 8]
```diff
@@ -0,0 +1,81 @@
+{ lib
+, bash
+, fetchFromGitHub
+, installShellFiles
+, makeWrapper
```
Comment (abathur) [inline pkgs/applications/system/zxfer/default.nix:5]:
```suggestion
```

no longer needed, yeah?

File: pkgs/applications/system/zxfer/default.nix [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  buildPhase = ''
+    substituteInPlace zxfer \
+      --replace 'AWK=$( which awk )'     'AWK=${gawk}/bin/awk' \
+      --replace 'AWK=$( which gawk )'    'AWK=${gawk}/bin/awk' \
+      --replace 'LCAT=""'                'LCAT=${coreutils}/bin/cat' \
+      --replace 'LZFS=$( which zfs )'    'LZFS=${zfs}/bin/zfs' \
+      --replace 'RSYNC=$( which rsync )' 'RSYNC=${rsync}/bin/rsync'
+  '';
+
+  nativeBuildInputs = [ installShellFiles makeWrapper ];
```
Comment (abathur) [inline pkgs/applications/system/zxfer/default.nix:36]:
```suggestion
  nativeBuildInputs = [ installShellFiles ];
```

Ditto :)

File: pkgs/applications/system/zxfer/default.nix [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+    installManPage zxfer.1m zxfer.8
+    install -Dm755 zxfer -t $out/bin/
+
+    runHook postInstall
+  '';
+
+  solutions.default = {
+    scripts = [ "bin/zxfer" ];
+    interpreter = "${bash}/bin/bash";
```
Comment (abathur) [inline pkgs/applications/system/zxfer/default.nix:49]:
```suggestion
    interpreter = "${bash}/bin/sh";
```

This one also uses a /bin/sh shebang https://github.com/allanjude/zxfer/blob/master/zxfer#L1

File: pkgs/applications/system/zxfer/default.nix [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    inputs = [
+      coreutils
+      gnugrep
+      gnused
+      openssh
+    ];
+    fake.external = [
+      "kldload"
+      "kldstat"
+      "svcadm"
```
Comment (abathur) [inline pkgs/applications/system/zxfer/default.nix:59]: Maybe good to note why these aren't available in a way that helps clarify if they're actionable later?
- Security wrappers? 
- Not packaged yet?
- Unpackageble? (proprietary system commands? etc.)

**Review by abathur — CHANGES_REQUESTED** [review]
nixpkgs-review was clean on x86_64-linux (NixOS) but I'll pick a few nits :)

File: pkgs/applications/system/zxfer/default.nix (hunk 11 above)
Comment (urandom2) [reply pkgs/applications/system/zxfer/default.nix]: yep, more investigation showed that they are builtins from other unixes. I have added comments to reflect their origin.

debian tried packaging kldutils, but seem to have given up, and I am unclear where opensolaris source lives.

File: pkgs/applications/system/zxfer/default.nix (hunk 7 above)
Comment (AndersonTorres) [reply pkgs/applications/system/zxfer/default.nix:21]: How many times this attribute will change? One or two in a decade, I suppose? This is too few to be afraid of typos.

But whatever.

File: pkgs/applications/system/zxfer/default.nix [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+  src = fetchFromGitHub {
+    owner = "allanjude";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-11SQJcD3GqPYBIgaycyKkc62/diVKPuuj2Or97j+NZY=";
+  };
+
+  buildPhase = ''
+    substituteInPlace zxfer \
```
Comment (AndersonTorres) [inline pkgs/applications/system/zxfer/default.nix:27]: This is literally a patch, not a build.

File: pkgs/applications/system/zxfer/default.nix [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+      --replace 'AWK=$( which awk )'     'AWK=${gawk}/bin/awk' \
```
Comment (AndersonTorres) [inline pkgs/applications/system/zxfer/default.nix:28]: Why not to put all of it as a buildInput plus a convenient makeWrapper?

File: pkgs/applications/system/zxfer/default.nix [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+  installPhase = ''
+    runHook preInstall
+
+    installManPage zxfer.1m zxfer.8
+    install -Dm755 zxfer -t $out/bin/
+
+    runHook postInstall
+  '';
+
+  solutions.default = {
```
Comment (AndersonTorres) [inline pkgs/applications/system/zxfer/default.nix:46]: what is this?

File: pkgs/applications/system/zxfer/default.nix [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+  buildPhase = ''
+    substituteInPlace zxfer \
+      --replace 'AWK=$( which awk )'     'AWK=${gawk}/bin/awk' \
+      --replace 'AWK=$( which gawk )'    'AWK=${gawk}/bin/awk' \
+      --replace 'LCAT=""'                'LCAT=${coreutils}/bin/cat' \
+      --replace 'LZFS=$( which zfs )'    'LZFS=${zfs}/bin/zfs' \
+      --replace 'RSYNC=$( which rsync )' 'RSYNC=${rsync}/bin/rsync'
+  '';
+
+  nativeBuildInputs = [ installShellFiles ];
```
Comment (AndersonTorres) [inline pkgs/applications/system/zxfer/default.nix:35]: This should be right after `src`.

**Review by AndersonTorres — DISMISSED** [review]

File: pkgs/applications/system/zxfer/default.nix (hunk 14 above)
Comment (abathur) [reply pkgs/applications/system/zxfer/default.nix:46]: Directives for resholve. https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/misc/resholve/README.md

**Review by abathur — APPROVED** [review]
This looks _plausible_ :)

(The parts that are within my knowledge LGTM; it looks complex otherwise, but I can tell you've wrestled with stuff like the local/remote behavior.)

**Pull request merged**